File: sdks/python/src/ai_provenance_protocol/_hash.py
"""Content hashing utilities."""
from __future__ import annotations

import hashlib
import json
from typing import Any, Dict

from ._types import AppMetadata, ContentHash


def canonicalize(content: Any) -> str:
    """Produce a canonical JSON representation of *content*.

    Keys are sorted and whitespace is stripped so the same logical object
    always produces the same byte string regardless of insertion order.

    Args:
        content: Any JSON-serialisable value.

    Returns:
        A compact, deterministic JSON string.
    """
    return json.dumps(content, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def hash_content(content: Any, *, algorithm: str = "sha256") -> ContentHash:
    """Compute a SHA-256 content hash of the canonical JSON representation.

    Args:
        content: Any JSON-serialisable value.
        algorithm: Hash algorithm identifier (currently only ``"sha256"`` is
            supported by the APP v1.0 specification).

    Returns:
        A :class:`ContentHash` dict with ``algorithm`` and ``value`` fields.

    Example::

        h = APP.hash_content({"reply": "Hello!"})
        # h == {"algorithm": "sha256", "value": "<hex>"}
    """
    if algorithm != "sha256":
        raise ValueError(f"Unsupported algorithm '{algorithm}'. Only 'sha256' is supported.")

    canonical = canonicalize(content)
    digest = hashlib.sha256(canonical.encode("utf-8")).hexdigest()
    return {"algorithm": algorithm, "value": digest}


def attach_hash(metadata: AppMetadata, content: Any) -> AppMetadata:
    """Compute and attach a content hash to an existing metadata record.

    Returns a *new* metadata dict with the ``content_hash`` field populated.

    Args:
        metadata: An existing APP metadata record.
        content: The content to hash (``_ai_provenance`` key is excluded if
            *content* is a dict).

    Returns:
        A new AppMetadata dict with ``content_hash`` set.
    """
    if isinstance(content, dict):
        content = {k: v for k, v in content.items() if k != "_ai_provenance"}

    return {**metadata, "content_hash": hash_content(content)}  # type: ignore[misc]

File: sdks/typescript/src/hash.ts
import { createHash } from 'node:crypto'
import type { HashAlgorithm } from './types.js'

/**
 * Produce the canonical string representation of a JSON object for hashing.
 * Removes the `_ai_provenance` key and sorts all object keys recursively.
 */
export function canonicalize(content: object): string {
  return JSON.stringify(sortKeys(removeAppKey(content)))
}

/**
 * Compute a content hash for integrity verification.
 * For JSON objects, uses canonical form (sorted keys, no _ai_provenance key).
 * For strings, hashes the raw UTF-8 bytes.
 *
 * Returns a string in the format `{algorithm}:{hex}`.
 */
export function hashContent(
  content: object | string,
  algorithm: HashAlgorithm = 'sha256'
): string {
  const data = typeof content === 'string' ? content : canonicalize(content)
  const hex = createHash(algorithm).update(data, 'utf8').digest('hex')
  return `${algorithm}:${hex}`
}

function removeAppKey(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(removeAppKey)
  if (value !== null && typeof value === 'object') {
    const result: Record<string, unknown> = {}
    for (const [k, v] of Object.entries(value)) {
      if (k !== '_ai_provenance') result[k] = removeAppKey(v)
    }
    return result
  }
  return value
}

function sortKeys(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(sortKeys)
  if (value !== null && typeof value === 'object') {
    return Object.fromEntries(
      Object.entries(value as object)
        .sort(([a], [b]) => a.localeCompare(b))
        .map(([k, v]) => [k, sortKeys(v)])
    )
  }
  return value
}
